### PR TITLE
Feature: Adding ▲ZEIT now deployment

### PR DESCRIPTION
**Pull request by lucaperret**
Deploy on [now](https://zeit.co/now) within Travis CI.

Here's the deployment: https://travis-ci.org/lucaperret/travis-now-provider/builds/359018542

### CODE REVIEW

File: lib/dpl/provider/now.rb [hunk 1]
```diff
@@ -0,0 +1,136 @@
+module DPL
+    class Provider
+      class Now < Provider
+        npm_g 'now'
+  
+        def directory
+          File.expand_path( (context.env['TRAVIS_BUILD_DIR'] || '.' ) + '/' + (options[:directory] || '') )
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: `TRAVIS_BUILD_DIR` should be defined in all our builds. If it is not, then there is something else that's grossly concerning.

`options.fetch(:directory, '')` makes the intent clearer.

File: lib/dpl/provider/now.rb [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        def deploy_name
+          options[:deploy_name]
+        end
+  
+        def team
+          if options[:team]
+            log "> Adding custom team scope #{options[:team]}"
+            now_team_option = "--team #{options[:team]}"
+          else
+            now_team_option = ''
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Initialize `new_team_option` to `''` at the top of the method, then drop this line.

File: lib/dpl/provider/now.rb [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            log '> No custom team scope provided.'
+          end
+
+          return now_team_option
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Explicit `return`s seem inconsistent with other parts of this class, so let's drop them.

File: lib/dpl/provider/now.rb [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        def rules_file
+          options[:rules_file]
+        end
+  
+        def token
+          context.env['NOW_TOKEN']
+        end
+
+        def auth
+          return "--token #{token} #{team}"
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: This is a bit strange. The invocation of `#team` here interlaces the log messages defined in that method. As `auth` is called from a few spots, I expect it to be quite noisy.

The credentials should be cached and logged at most once.

File: lib/dpl/provider/now.rb [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+          end
+          return deployment_url
+        end
+
+        def check_auth
+          if ! token then raise Error, '> Error!! Please add NOW_TOKEN Environment Variables in Travis settings (get your token here https://zeit.co/account/tokens)' end
+        end
+  
+        def check_app
+          if ! File.directory?(directory) then raise Error, 'Please set a valid project folder path in .travis.yml under deploy: directory: myPath' end
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: This one liner seems a bit too wide. Please break it up into shorter lines. Also, I personally prefer `unless` here to `if !`.

File: lib/dpl/provider/now.rb [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+          if cleanup && !alias_url then raise Error, 'You must set the alias parameter when using the cleanup parameter so that now.sh knows which deployments to remove!' end
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: This should be broken up into shorter lines as well.

File: lib/dpl/provider/now.rb [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+          if ((rules_domain && !rules_file) || rules_file && !rules_domain) then raise Error, 'You must set the rules_domain and rules_file parameters in order to set custom domain rules' end
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Ditto.

File: lib/dpl/provider/now.rb [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        end
+  
+        def needs_key?
+          false
+        end
+  
+        def push_app
+          log "> Deploying #{directory} on now.sh..."
+          deployment_url = `now #{auth} #{deploy_options} #{directory}`
+          log "> Success! Deployment complete to #{deployment_url}"
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Is this true? The back ticks just returns `STDOUT`, regardless of what the command saw. In other words, `deployment_url` will be what the `now` command writes to `STDOUT` regardless of the result of the deployment. The provider needs to be more intelligent about the result of the deployment.

File: lib/dpl/provider/now.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            log scale_success_message
+          end
+
+          if rules_domain
+            log '> Assigning domain rules...'
+            rules_success_message = `now alias #{auth} #{rules_domain} -r #{rules_file}`
+            log rules_success_message
+          end
+
+          if $?.exitstatus != 0
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: I am not sure if `$?` has what you think it does. The only `system` call I can find is in `#aliasing`, and it runs only when `alias` option is set.

File: lib/dpl/provider/now.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+        def push_app
+          log "> Deploying #{directory} on now.sh..."
+          deployment_url = `now #{auth} #{deploy_options} #{directory}`
+          log "> Success! Deployment complete to #{deployment_url}"
+
+          deployment_url = aliasing(deployment_url)
+
+          if cleanup
+            log '> Cleaning up old deployments...'
+            cleanup_success_message = `now rm --safe --yes #{auth} #{alias_url}`
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: It does not seem necessary to me to assign `cleanup_success_message` here. Simply call `#log`.

File: lib/dpl/provider/now.rb [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+            log cleanup_success_message
+          end
+
+          if scale
+            log '> Scaling...'
+            scale_success_message = `now scale #{auth} #{deployment_url} #{scale}`
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Same.

File: lib/dpl/provider/now.rb [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            log scale_success_message
+          end
+
+          if rules_domain
+            log '> Assigning domain rules...'
+            rules_success_message = `now alias #{auth} #{rules_domain} -r #{rules_file}`
```
Comment (BanzaiMan) [inline lib/dpl/provider/now.rb]: Same.

**Comment by BanzaiMan** [comment]
Sorry for the long delay. I've made some suggestions on this PR.

Thank you!

**Comment by svenfuchs** [comment]
Hey @lucaperret :wave:

We are working on releasing a new major version dpl v2. The respective branch has been merged to master, and we are planning on releasing a developer preview, as well as a blog post soon, hopefully later this week.

I have ported your work on this branch to a new PR here: https://github.com/travis-ci/dpl/pull/1051. I believe this should be equivalent to your work.

If you would like me to amend the commit in order to add your Git user name and email address please let me know. (My commit was based on your work, so I would happily pass authorship for it, but I also did not want to presume.)

It would be great if you could help us test if the issue that your change was meant to address. You can opt into using dpl v2 by adding this to your `.travis.yml` file:

```yaml
deploy:
  provider: [provider-name]
  edge:
    branch: master
⋮
```

Thank you!

**Comment by svenfuchs** [comment]
i am going to close this PR here in favor of https://github.com/travis-ci/dpl/pull/1051. please feel free to reopen if needed.

**Pull request closed without merging**